source/server/admin/base_stats_request.h

```diff
@@ -0,0 +1,147 @@
+  //
+  // For text/JSON/HTML stats, we do 3 passes through all the scopes_, and we emit
+  // text-readouts first, then the intermingled counters and gauges, and finally
+  // the histograms. For prometheus stats, we do 3 or 4 passes (this depends on whether Text
+  // Readouts are explicitly requested) through all the scopes_, and we emit counters first,
+  // then gauges, possibly text readouts and finally histograms.
+  bool nextChunk(Buffer::Instance& response) override;
+
+  // To duplicate prior behavior for this class, we do several passes over all the stats. For
+  // text/JSON/HTML stats, we do 3 passes:
+  //   1. text readouts across all scopes
+  //   2. counters and gauges, co-mingled, across all scopes
+  //   3. histograms across all scopes.
+  // For prometheus stats, we do 3 or 4 passes:
+  //   1. counters across all scopes
+  //   2. gauges across all scopes
+  //   3. text readouts (only if explicitly requested via query param)
+  //   4. histograms across all scopes.
+  // It would be little more efficient to co-mingle all the stats, but three
+  // passes over the scopes is OK. In the future we may decide to organize the
+  // result data differently, but in the process of changing from buffering
+  // the entire /stats response to streaming the data out in chunks, it's easier
+  // to reason about if the tests don't change their expectations.
+  void startPhase();
+
+  // Iterates over scope_vec and populates the metric types associated with the
+  // current phase.
+  void populateStatsForCurrentPhase(const ScopeVec& scope_vec);
+
+  virtual Stats::IterateFn<Stats::TextReadout> checkStatForTextReadout() PURE;
+  virtual Stats::IterateFn<Stats::Gauge> checkStatForGauge() PURE;
+  virtual Stats::IterateFn<Stats::Counter> checkStatForCounter() PURE;
+  virtual Stats::IterateFn<Stats::Histogram> checkStatForHistogram() PURE;
+
+  virtual void processTextReadout(const std::string& name, Buffer::Instance& response,
+                                  const StatOrScopes& variant) PURE;
+  virtual void processGauge(const std::string& name, Buffer::Instance& response,
+                            const StatOrScopes& variant) PURE;
+  virtual void processCounter(const std::string& name, Buffer::Instance& response,
+                              const StatOrScopes& variant) PURE;
+  virtual void processHistogram(const std::string& name, Buffer::Instance& response,
+                                const StatOrScopes& variant) PURE;
+
+  // Sets the chunk size.
+  void setChunkSize(uint64_t chunk_size) { chunk_size_ = chunk_size; }
+
+protected:
+  StatsParams params_;
+  Stats::Store& stats_;
+  ScopeVec scopes_;
+  absl::btree_map<std::string, StatOrScopes> stat_map_;
+  std::unique_ptr<StatsRender> render_;
+  Buffer::OwnedImpl response_;
+  UrlHandlerFn url_handler_fn_;
+  uint64_t chunk_size_{DefaultChunkSize};
+
+  // phase-related state
+  uint64_t phase_stat_count_{0};
+  unsigned short phase_;
```

Comment:
The index doesn't have to do with skipping the text-readouts per se. It's used to keep track of the current phase in the shared base class, and move on to the next phase w/o having to know what that phase is.

This is because the prom and non-prom algorithms go through phases in a different order. I hope that makes sense: if you can think of a better way to achieve something similar, I am happy to update the code. For now I will rename the field name and type.